File: pnpm11/installing/deps-resolver/src/depPathCompatibility.ts
import type { DepPath } from '@pnpm/types'

import type {
  GenericDependenciesGraphNodeWithResolvedChildren,
  GenericDependenciesGraphWithResolvedChildren,
  PartialResolvedPackage,
} from './resolvePeers.js'

// Shared helpers used by both resolvePeers (dedupePeerDependents) and
// dedupeInjectedDeps. Lives in its own module so neither consumer has to import
// the other, which would create a runtime cycle. The type imports above are
// erased at build time, so no cycle exists at runtime.

export function nodeDepsCount (node: GenericDependenciesGraphNodeWithResolvedChildren): number {
  return Object.keys(node.children!).length + node.resolvedPeerNames.size
}

// Whether `depPath1` is a superset-or-equal of `depPath2`: same-or-more resolved
// children and peers. Each child alias of `depPath2` must resolve either to the
// same depPath or to a variant of the same package that is itself a
// superset-or-equal, which is what lets a package whose child carries a peer
// suffix absorb the variant whose child does not.
// Compares dependency/peer *sets* only, not package identity, so callers must
// pass depPaths already known to share a `pkgIdWithPatchHash` — otherwise two
// unrelated leaf packages (both with empty sets) would count as compatible.
export function isCompatibleAndHasMoreDeps<T extends PartialResolvedPackage> (
  depGraph: GenericDependenciesGraphWithResolvedChildren<T>,
  depPath1: DepPath,
  depPath2: DepPath
): boolean {
  // Pairs are recorded on entry, so a dependency cycle assumes compatibility
  // instead of recursing forever. The assumption is never observed by a `true`
  // result: an incompatible pair anywhere aborts the whole check.
  const visited = new Map<DepPath, Set<DepPath>>()
  const isSuperset = (supersetDepPath: DepPath, subsetDepPath: DepPath): boolean => {
    if (supersetDepPath === subsetDepPath) return true
    let subsets = visited.get(supersetDepPath)
    if (subsets == null) {
      subsets = new Set()
      visited.set(supersetDepPath, subsets)
    }
    if (subsets.has(subsetDepPath)) return true
    subsets.add(subsetDepPath)

    const supersetNode = depGraph[supersetDepPath]
    const subsetNode = depGraph[subsetDepPath]
    if (supersetNode == null || subsetNode == null) return false
    if (nodeDepsCount(supersetNode) < nodeDepsCount(subsetNode)) return false

    for (const peerName of subsetNode.resolvedPeerNames) {
      if (!supersetNode.resolvedPeerNames.has(peerName)) return false
    }

    return Object.entries(subsetNode.children!).every(([alias, subsetChildDepPath]) => {
      const supersetChildDepPath = supersetNode.children![alias]
      if (supersetChildDepPath == null) return false
      if (supersetChildDepPath === subsetChildDepPath) return true
      const supersetChildNode = depGraph[supersetChildDepPath]
      const subsetChildNode = depGraph[subsetChildDepPath]
      return supersetChildNode != null &&
        subsetChildNode != null &&
        supersetChildNode.pkgIdWithPatchHash === subsetChildNode.pkgIdWithPatchHash &&
        isSuperset(supersetChildDepPath, subsetChildDepPath)
    })
  }
  return isSuperset(depPath1, depPath2)
}

File: pnpm/crates/resolving-deps-resolver/src/dep_path_compatibility.rs
//! Shared depPath-compatibility helpers used by both
//! [`fn@crate::dedupe_peer_dependents::dedupe_peer_dependents`] and
//! [`fn@crate::dedupe_injected_deps::dedupe_injected_deps`].

use std::collections::VecDeque;

use pnpm_deps_path::DepPath;
use rustc_hash::FxHashSet as HashSet;

use crate::dependencies_graph::{DependenciesGraph, DependenciesGraphNode};

/// Number of edges a variant carries: its child dependencies plus the
/// peers it resolved against its ancestors.
pub(crate) fn node_deps_count(node: &DependenciesGraphNode) -> usize {
    node.edges.children.len() + node.edges.resolved_peer_names.len()
}

/// Whether `larger` can absorb `smaller`: it must have at least as many
/// deps, every peer `smaller` resolved must also be resolved by `larger`,
/// and each of `smaller`'s child aliases must resolve either to the same
/// depPath or to a variant of the same package that `larger`'s child can
/// absorb in turn. That last case is what lets a package whose child
/// carries a peer suffix absorb the variant whose child does not.
///
/// Compares dependency/peer *sets* only, not package identity, so callers
/// must pass depPaths already known to share a `pkgIdWithPatchHash` —
/// otherwise two unrelated leaf packages (both with empty sets) would
/// count as compatible.
///
/// A pair reached twice is taken as compatible, which is what terminates
/// dependency cycles.
pub(crate) fn is_compatible_and_has_more_deps<'a>(
    graph: &'a DependenciesGraph,
    larger: &'a DepPath,
    smaller: &'a DepPath,
) -> bool {
    let mut queue = PairQueue::new(larger, smaller);
    while let Some((larger, smaller)) = queue.pop() {
        if !pair_is_compatible(graph, larger, smaller, &mut queue) {
            return false;
        }
    }
    true
}

/// Pairs still to check, breadth-first, so the shallowest incompatible
/// pair settles the whole check. A pair is recorded when it enters, so an
/// edge that reaches one already queued adds nothing.
struct PairQueue<'a> {
    pending: VecDeque<(&'a DepPath, &'a DepPath)>,
    queued: HashSet<(&'a DepPath, &'a DepPath)>,
}

impl<'a> PairQueue<'a> {
    fn new(larger: &'a DepPath, smaller: &'a DepPath) -> Self {
        let mut queue = PairQueue { pending: VecDeque::new(), queued: HashSet::default() };
        queue.push(larger, smaller);
        queue
    }

    fn push(&mut self, larger: &'a DepPath, smaller: &'a DepPath) {
        if larger != smaller && self.queued.insert((larger, smaller)) {
            self.pending.push_back((larger, smaller));
        }
    }

    fn pop(&mut self) -> Option<(&'a DepPath, &'a DepPath)> {
        self.pending.pop_front()
    }
}

/// Checks one pair and queues the child pairs whose depPaths differ.
fn pair_is_compatible<'a>(
    graph: &'a DependenciesGraph,
    larger: &DepPath,
    smaller: &DepPath,
    queue: &mut PairQueue<'a>,
) -> bool {
    let (Some(larger_node), Some(smaller_node)) = (graph.get(larger), graph.get(smaller)) else {
        return false;
    };

    if node_deps_count(larger_node) < node_deps_count(smaller_node) {
        return false;
    }

    if !has_all_resolved_peers(larger_node, smaller_node) {
        return false;
    }

    smaller_node.edges.children
        .iter()
        .all(|(alias, smaller_child)| {
            queue_child_pair(graph, larger_node, alias, smaller_child, queue)
        })
}

fn has_all_resolved_peers(
    larger_node: &DependenciesGraphNode,
    smaller_node: &DependenciesGraphNode,
) -> bool {
    smaller_node.edges.resolved_peer_names
        .iter()
        .all(|peer| larger_node.edges.resolved_peer_names.contains(peer))
}

fn queue_child_pair<'a>(
    graph: &'a DependenciesGraph,
    larger_node: &'a DependenciesGraphNode,
    alias: &str,
    smaller_child: &'a DepPath,
    queue: &mut PairQueue<'a>,
) -> bool {
    let Some(larger_child) = larger_node.edges.children.get(alias) else {
        return false;
    };
    if larger_child == smaller_child {
        return true;
    }
    let (Some(larger_child_node), Some(smaller_child_node)) =
        (graph.get(larger_child), graph.get(smaller_child))
    else {
        return false;
    };
    if larger_child_node.resolved_package_id != smaller_child_node.resolved_package_id {
        return false;
    }
    queue.push(larger_child, smaller_child);
    true
}
